File: python/Q778_swim_in_rising_water.py
#
# @lc app=leetcode id=778 lang=python3
#
# [778] Swim in Rising Water
#

# @lc code=start
import heapq


class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        ROWS: int = len(grid)
        COLS: int = len(grid[0])
        directions: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        heap: list[tuple[int, int, int]] = [(grid[0][0], 0, 0)]
        visited: set[tuple[int, int]] = set()

        while heap:
            height, row, col = heapq.heappop(heap)

            if row == ROWS - 1 and col == COLS - 1:
                return height

            for x, y in directions:
                new_row: int = row + x
                new_col: int = col + y

                if (
                    0 <= new_row < ROWS
                    and 0 <= new_col < COLS
                    and (new_row, new_col) not in visited
                ):
                    visited.add((new_row, new_col))

                    if height > grid[new_row][new_col]:
                        heapq.heappush(heap, (height, new_row, new_col))
                    else:
                        heapq.heappush(heap, (grid[new_row][new_col], new_row, new_col))

        return -1


# @lc code=end

File: rust/src/leetcode/Q778_swim_in_rising_water.rs
/*
 * @lc app=leetcode id=778 lang=rust
 *
 * [778] Swim in Rising Water
 */

// @lc code=start
use std::collections::BinaryHeap;

impl Solution {
    pub fn swim_in_water(grid: Vec<Vec<i32>>) -> i32 {
        let size: usize = grid.len();
        static DIRECTIONS: [(i32, i32); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];

        let mut visited: Vec<bool> = vec![false; size * size];
        let mut explore: BinaryHeap<(i32, i32, i32)> = BinaryHeap::new();

        explore.push((-grid[0][0], 0, 0));

        while !explore.is_empty() {
            let (time, row, col) = explore.pop().unwrap();

            if visited[row as usize * size + col as usize] {
                continue;
            }

            if row == size as i32 - 1 && col == size as i32 - 1 {
                return -time;
            }

            visited[row as usize * size + col as usize] = true;

            for (dx, dy) in DIRECTIONS {
                let new_row: i32 = row + dx;
                let new_col: i32 = col + dy;

                if new_row >= 0
                    && new_col >= 0
                    && new_row < size as i32
                    && new_col < size as i32
                    && !visited[new_row as usize * size + new_col as usize]
                {
                    explore.push((
                        time.min(-grid[new_row as usize][new_col as usize]),
                        new_row,
                        new_col,
                    ));
                }
            }
        }

        -1
    }
}
// @lc code=end
